### PR TITLE
Load z3 checkpoints for inference

### LINKED ISSUE #2449: [BUG] stage 3 cannot load the checkpoint when optimizer is not configured
**Describe the bug**
Deepspeed stage 3 cannot load the checkpoint when optimizer is not configured during inferencing.

Error: `'DeepSpeedZeRoOffload' object has no attribute 'checkpoint_event_prologue'`

I used a very basic pytorch-lightning script to generate the issue. I am a maintainer at lightning, so feel free to ping me for any information.
```py
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        accelerator='gpu',
        devices=2,
        strategy='deepspeed_stage_3'
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.save_checkpoint('deepspeed_ckpt')
    trainer.test(model, dataloaders=test_data, ckpt_path='deepspeed_ckpt')


if __name__ == "__main__":
    run()

```

**To Reproduce**
1. Run the above script
2. deepspeed==0.7.4, torch==1.12.1+cu116
3. python script_name.py

**Expected behavior**
It should work without any error since the engine is reinitilaized for inference without any optimizer as mentioned here: https://deepspeed.readthedocs.io/en/latest/model-checkpointing.html#deepspeed.DeepSpeedEngine.load_checkpoint

**ds_report output**
Please run `ds_report` to give us details about your setup.

```console
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
sparse_attn ............ [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
utils .................. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/home/rohit/miniconda3/envs/pl/lib/python3.9/site-packages/torch']
torch version .................... 1.12.1+cu116
torch cuda version ............... 11.6
torch hip version ................ None
nvcc version ..................... 11.6
deepspeed install path ........... ['/home/rohit/miniconda3/envs/pl/lib/python3.9/site-packages/deepspeed']
deepspeed info ................... 0.7.4, unknown, unknown
deepspeed wheel compiled w. ...... torch 1.12, cuda 11.6
```

**Screenshots**

**System info (please complete the following information):**
 - OS: [e.g. Ubuntu 18.04]
 - GPU count and types [e.g. two machines with x8 A100s each]
 - (if applicable) what [DeepSpeed-MII](https://github.com/microsoft/deepspeed-mii) version are you using
 - (if applicable) Hugging Face Transformers/Accelerate/etc. versions
 - Python version
 - Any other relevant info about your setup

**Docker context**
Are you using a specific docker image that you can share?

**Additional context**

### LINKED ISSUE #3601: AttributeError: 'DeepSpeedZeRoOffload' object has no attribute 'checkpoint_event_prologue' [BUG]
during inference, the error "AttributeError: 'DeepSpeedZeRoOffload' object has no attribute 'checkpoint_event_prologue'" occur, similar to #2449.

**To Reproduce**

```
train_params = dict(
                strategy = DeepSpeedStrategy(
        stage=3,
        offload_optimizer=args.cpu_offload,
        offload_parameters=args.cpu_offload,
    ) if args.strategy == 'deepspeed_stage_3' else args.strategy,
                accumulate_grad_batches=args.gradient_accumulation_steps,
                accelerator='gpu' if args.n_gpu > 0 else None,
                devices = [args.gpu_no] if args.n_gpu <= 1 else args.n_gpu,
                # num_nodes = args.n_gpu,
                max_epochs=args.num_train_epochs,
                precision='bf16' if args.fp_16 else 32,
                gradient_clip_val=args.max_grad_norm, # default args.max_grad_norm
                enable_checkpointing =False,
                callbacks=callbacks,
                # log_every_n_steps = sum([len(t[0]) for t in curr_train_ids])//args.train_batch_size
                )
    trainer = pl.Trainer(**train_params)
    ckpt_path = os.path.join(args.output_dir,'best.ckpt')
    
    
    # Train model
    trainer.fit(model)
    trainer.save_checkpoint(ckpt_path)
    ## Test phase
    test_metrics = trainer.test() 
```error happened here, the trainer loads the ckpt_path saved in the logs and somehow has that error.

I was able to mitigate it by setting ckpt_path to 'last' but it started printing an extremely long list of weird numbers like
`1017, 1024, 1018, 1019, 1023, 1020, 1022, 1021, 1025, 1026, 1028, 1034, 1029, 1030, 1031, 1032, 1033, 1035, 1036, 1042, 1037, 1038, 1041, 1043, 1044, 1051, 1045, 1046, 1050, 
1047, 1049, 1048, 1052, 1053, 1055, 1061, 1056, 1057, 1058, 1059, 1060, 1062, 1063, 1069, 1064, 1065, 1068, 1070, 1071, 1078, 1072, 1073, 1077, 1074, 1076, 1075, 1079, 1080, 
1082, 1088, 1083, 1084, 1085, 1086, 1087, 1089, 1090, 1096, 1091, 1092, 1095, 1097, 1098, 1105, 1099, 1100, 1104, 1101, 1103, 1102, 1106, 1107, 1109, 1115, 1110, 1111, 1112, 
1113, 1114, 1116, 1117, 1123, 1118, 1119, 1122, 1124, 1125, 1132, 1126, 1127, 1131, 1128, 1130, 1129, 1133, 1134, 1135, 1136]`

**ds_report output**
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
async_io ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  using untested triton version (2.0.0), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
utils .................. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/home/weijie/anaconda3/envs/flan/lib/python3.10/site-packages/torch']
torch version .................... 1.12.1+cu113
deepspeed install path ........... ['/home/weijie/anaconda3/envs/flan/lib/python3.10/site-packages/deepspeed']
deepspeed info ................... 0.8.2, unknown, unknown
torch cuda version ............... 11.3
torch hip version ................ None
nvcc version ..................... 11.7
deepspeed wheel compiled w. ...... torch 1.12, cuda 11.3

### PR DESCRIPTION
**Pull request by tjruwase**
Fix #2449
Fix #3601

### CODE REVIEW

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -2686,6 +2693,12 @@ def _load_checkpoint(self,
         if checkpoint is None:
             return None, None
 
+        fetch_z3_params = False
+        if self.zero_optimization_partition_weights() and not load_optimizer_states:
+            from deepspeed.utils.zero_to_fp32 import get_fp32_state_dict_from_zero_checkpoint
```
Comment (mrwyattii) [inline deepspeed/runtime/engine.py:2698]: Any reason to have this import here instead of at the top?

File: deepspeed/runtime/engine.py [hunk 2]
```diff
@@ -2641,7 +2647,7 @@ def load_checkpoint(self,
                     )
                     return None, None
 
-        if self.zero_optimization_partition_weights():
+        if self.optimizer and hasattr(self.optimizer, 'checkpoint_event_prologue'):
```
Comment (mrwyattii) [inline deepspeed/runtime/engine.py:2650]: I see this repeated several times in the added code. Perhaps create a function like `self.optimizer_has_ckpt_event_prologue()`?

File: deepspeed/runtime/engine.py [hunk 3]
```diff
@@ -2503,14 +2503,20 @@ def load_moe_state_dict(checkpoint_path,
...
-        else:
-            self.module.load_state_dict(
-                module_state_dict,  # TODO
-                strict=strict)
+    def load_module_state_dict(self, checkpoint, strict=True, custom_load_fn=None, fetch_z3_params=False):
+
+        def z3_params_to_fetch(param_list):
+            return [p for p in param_list if hasattr(p, 'ds_id') and p.ds_status == ZeroParamStatus.NOT_AVAILABLE]
+
+        params_to_fetch = z3_params_to_fetch(self.module.parameters()) if fetch_z3_params else []
```
Comment (mrwyattii) [inline deepspeed/runtime/engine.py:2511]: Ternary conditions are often more difficult to read, potentially better alternative:

```suggestion
        def z3_params_to_fetch(param_list):
            if not fetch_z3_params:
                return []
            return [p for p in param_list if hasattr(p, 'ds_id') and p.ds_status == ZeroParamStatus.NOT_AVAILABLE]

        params_to_fetch = z3_params_to_fetch(self.module.parameters())
```

File: deepspeed/runtime/engine.py (hunk 1 above)
Comment (tjruwase) [reply deepspeed/runtime/engine.py:2698]: No good reason. Will move to the top.

File: deepspeed/runtime/engine.py (hunk 2 above)
Comment (tjruwase) [reply deepspeed/runtime/engine.py:2650]: Good point

File: deepspeed/runtime/engine.py (hunk 3 above)
Comment (tjruwase) [reply deepspeed/runtime/engine.py:2511]: I will rewrite differently.

**Review by mrwyattii — APPROVED** [review]

**Pull request merged**